### DISCUSSION TITLE
Can you disable the snippets?

### DISCUSSION THREAD

**Discussion by tymekm** [clangd/clangd#1867] [Q&A] [answered]
Is there a way to disable the snippets shipped with clang e.g.
```
for (init-statement; condition; inc-expression) {statements}~
```
while keeping the snippet behavior of adding parenthesis and filling in the arguments list when completing a function?

Thanks

**Comment by HighCommander4** [comment, marked as answer]
It doesn't look like there is currently a way to do this.

If you're interested in adding one, I'm happy to provide guidance.

**Reply by tymekm** [reply]
Thank you for your reply. I'd like to add that functionality if you can point me in the right direction.

**Reply by HighCommander4** [reply]
Sure. The snippets of the sort you're talking about are referred to in the LLVM codebase as "code patterns".

The place we'll ultimately want to change to respect our new option is [this line](https://searchfox.org/llvm/rev/b6cce87110072a2db19276e042cd40b06285abbc/clang-tools-extra/clangd/CodeComplete.cpp#2102).

The existing `EnableSnippets` variable is determined by whether the language client supports snippets as a type of code completion item, and applies to both code patterns and function argument placeholders. So we'll want to change the line to assign `EnableSnippets && OurNewOption` instead of just `EnableSnippets`.

Regarding the form our new option will take, I recommend a config key under [Completion](https://clangd.llvm.org/config.html#completion) in the `.clangd` config file, so that you'd disable it like so:

```yaml
Completion:
  CodePatterns: false
```

There is some plumbing involved in adding a new config option. I recommend looking at a previous patch that added a boolean config option (for example [this one](https://reviews.llvm.org/D116713), which added the ability to enable/disable inlay hints), and looking at the various places that need modification.

If you have any questions, please feel free to reach out here or on [Discord](https://discord.gg/xS7Z362).

**Reply by Noustaa** [reply]
Hello :) I'm currently implementing this in LLVM following your guidance.
It works fine to fix the issue described above=> "for (init-statement; condition; inc-expression) {statements}~" not showing anymore when adding this to config (.clangd):
```yaml
Completion:
    CodePatterns: false
```

Although, there is still plenty of others "Snippets" showing up:
![image](https://github.com/user-attachments/assets/82a54dca-aed1-4033-a4ee-1fe62f7d811e)
![image](https://github.com/user-attachments/assets/56e62c7f-79f6-4d44-8fa8-e9fd0bc9e979)

Any way to filter them aswell ?

**Reply by HighCommander4** [reply]
> Although, there is still plenty of others "Snippets" showing up: 

Interesting.

Looking at the code that produces these patterns, which is in the clang libraries, `namespace <identifier> { <declarations> }` [is conditional](https://searchfox.org/llvm/rev/6bb2f90557fb2b4b216299cc2beb4afb641476aa/clang/lib/Sema/SemaCodeComplete.cpp#2154-2166) on `includeCodePatterns()`, but `using namespace <identifier>` [is not](https://searchfox.org/llvm/rev/6bb2f90557fb2b4b216299cc2beb4afb641476aa/clang/lib/Sema/SemaCodeComplete.cpp#2177-2182).

It's not clear to me what the rationale for this distinction is.

> Any way to filter them aswell ?

A more aggressive form of filtering would be to skip any item in [this loop](https://searchfox.org/llvm/rev/6bb2f90557fb2b4b216299cc2beb4afb641476aa/clang-tools-extra/clangd/CodeComplete.cpp#952) with `Result.Kind == CodeCompletionResult::RK_Pattern`.

It's not clear to me though that this is what users would expect from an option to disable "snippets". For example, consider `static_cast<type>(expression)`. That is an `RK_Pattern` completion and so would be excluded, but it's an expression-context completion, which from a user's point of view it's pretty similar to, say `boost::lexical_cast<type>(expression)` (which is `RK_Declaration` and would remain) and just happens to involve a keyword rather than a function; is that really a "snippet" in a relevant sense?

**Reply by Noustaa** [reply]
I think the word "Snippet" is kind of overused in the clang codebase: From my understanding it is also referring to the fact that when you call a function, it will automatically add the placeholders in it so you can complete the arguments by tabbing from one to the other.

I don't know if my PR is enough right now as it is disabling only a few code patterns (AKA snippets).
Should i add several levels of aggressivity in filtering ?
I will take a look the the loop you pointed out to understand all the different types of patterns.

**Reply by Noustaa** [reply]
I confirm that skipping inside the loop with `Result.Kind == CodeCompletionResult::RK_Pattern` works as expected.
No snippets from LSP are showing up anymore.

What do you suggest then for the PR ?

EDIT: Actually, 14 tests broke with this change

**Reply by HighCommander4** [reply]
> Looking at the code that produces these patterns, which is in the clang libraries, `namespace <identifier> { <declarations> }` [is conditional](https://searchfox.org/llvm/rev/6bb2f90557fb2b4b216299cc2beb4afb641476aa/clang/lib/Sema/SemaCodeComplete.cpp#2154-2166) on `includeCodePatterns()`, but `using namespace <identifier>` [is not](https://searchfox.org/llvm/rev/6bb2f90557fb2b4b216299cc2beb4afb641476aa/clang/lib/Sema/SemaCodeComplete.cpp#2177-2182).
> 
> It's not clear to me what the rationale for this distinction is.

[This commit](https://github.com/llvm/llvm-project/commit/f4c33349b59795c9abb9a661eb0b6eae7b628f72) message provides a clue: "Make -code-completion-patterns only cover multi-line code". So, I guess it's only multi-line patterns which are guarded by `IncludeCodePatterns` setting.

Unfortunately, the commit in question does not have a linked issue, and dates back to 2010 when the project did not yet use Phabricator for code reviews, so I can't find any additional context about the motivation for the change.

**Reply by HighCommander4** [reply]
> What do you suggest then for the PR ?

Before making a suggestion for a PR, I'd like to understand your perspective as a user (and also @tymekm's as the original poster).

Given the following categories of code patterns:

 1. Patterns in expression context, e.g. `static_cast<type>(expression)`
 2. Single-line patterns in statement context, e.g. `using namespace <identifier>;`
 3. Multi-line patterns in statement context, e.g. 
    ```
    namespace <identifier> {
      <declarations>
    }
    ```

Which of these categories are you looking to disable, and why?

If the problem is that the patterns push other completions further down, could the problem be addressed by re-ordering rather than removing?

**Reply by Noustaa** [reply]
My main motivation was to get rid of this:
![image](https://github.com/user-attachments/assets/ff9c6070-ca5b-4913-b82c-099cee8da12b)
I do have some custom snippets (the one tagged "snippets" on the right) and in this case i get duplicates (.e.g they provide the same functionnality) with the ones from LSP (clangd).

I think LSP snippets are fine enabled by default for a user that won't use custom snippets but i would prefer to create all the snippets i need myself and be able to disable snippets from other sources (.e.g lsp).

> Which of these categories are you looking to disable, and why?

To answer: I would like to disable all of them because i already have my custom source of snippets.

**Reply by HighCommander4** [reply]
Ok, then I would suggest adding a config option like this:

```yaml
Completion:
  CodePatterns: None
```

which disables all the mentioned categories. The default value of the option can be named `All`.

By naming them `None` and `All` rather than `Disabled` and `Enabled`, we leave the possibility of other values open in the future, e.g. `SingleLineOnly` or whatever, but that can be left for a future patch if/when someone actually wants it.

**Reply by Noustaa** [reply]
Very good suggestion !
PR has been submitted: https://github.com/llvm/llvm-project/pull/137613

Thank you !
